### PR TITLE
Lazy-load MPI using Requires

### PR DESCRIPTION
**Pull request by jipolanco**
As discussed in #619, this proposal modifies parallel HDF5 support, defining certain MPI I/O specific functions only if MPI.jl was previously loaded. This is done using [Requires](https://github.com/JuliaPackaging/Requires.jl), so that MPI.jl is not added to the dependency list of HDF5.jl.

The main advantage of this approach is that we now have access to types such as `MPI.MPI_Comm` (a C object) and `MPI.Comm` (its Julia wrapper). In particular, we know their sizes, which simplifies the previous implementation, and solves some (minor) type inference issues.

MPI I/O functions such as `h5p_set_fapl_mpio` and `h5p_get_fapl_mpio` now have definitions that accept (or return) MPI wrappers, which are properly managed by MPI.jl. This greatly simplifies opening parallel HDF5 files, as `MPI.Comm` and `MPI.Info` objects no longer need to be converted to their HDF5 counterparts by the user.

In particular, the following is now possible with the proposed changes:

```julia
comm = MPI.COMM_WORLD
info = MPI.Info()

h5open("filename.h5", "w", "fapl_mpio", (comm, info)) do ff
    # do stuff...
end
```

In its current form, this proposal breaks compatibility with previous versions with regard to parallel HDF5 support, since the `mpihandles` dict and `h5_mpihandle` function were removed. If this is not acceptable, I guess these could be kept for backwards compatibility. It also breaks compatibility in a second way (very minor, in my opinion): due to Requires, MPI necessarily needs to be imported before HDF5 for parallel I/O to work.

Finally, this updates the `test/mpio.jl` script (which was broken under recent MPI.jl versions) to use the new functionality.

### CODE REVIEW

**Comment by musm** [comment]
We plan to make a breaking release soon.  Perhaps this should go on?

File: src/HDF5.jl [hunk 1]
```diff
@@ -2572,6 +2555,18 @@ const ASCII_ATTRIBUTE_PROPERTIES = Ref{HDF5Properties}()
 
 const DEFAULT_PROPERTIES = HDF5Properties(H5P_DEFAULT, false, H5P_DEFAULT)
 
+const HAS_PARALLEL = Ref(false)
```
Comment (musm) [inline src/HDF5.jl:2558]: Does this need to be a  constant ref?

File: src/mpio.jl [hunk 2]
```diff
@@ -0,0 +1,61 @@
+using .MPI
```
Comment (musm) [inline src/mpio.jl:1]:
```suggestion
using MPI
```
?

**Comment by musm** [comment]
I'm not too familiar with the MPI stuff, but seems reasonable. Perhaps some documentation would be in order as well.

**Comment by jipolanco** [comment]
Thanks for looking at this. I just rebased to the current master.

Tests don't pass yet, as I'm still working on adapting things to the changes in #632 (which are a great improvement!).

I'll also work on the docs.

File: src/mpio.jl (hunk 2 above)
Comment (jipolanco) [reply src/mpio.jl:1]: Should be `using .MPI`. That's just how Requires works.

From their [README](https://github.com/JuliaPackaging/Requires.jl/blob/master/README.md):

> In the @require block, or any included files, you can use or import the package, but note that you must use the syntax `using .Gadfly` or `import .Gadfly`, rather than the usual syntax. Otherwise you will get a warning about Gadfly not being in dependencies.

File: src/HDF5.jl (hunk 1 above)
Comment (jipolanco) [reply src/HDF5.jl:2565]: That value gets modified to `Ref(true)` if the `mpio.jl` file gets loaded via Requires, and if the HDF5 libraries include parallel support.

https://github.com/JuliaIO/HDF5.jl/blob/a12b66f9af28a76cbf7b099ed81709187744767f/src/mpio.jl#L4-L7

I thought it was a simple way to allow for this kind of dual behaviour (depending on whether code is loaded or not), but there may be a better way to do this.

File: src/HDF5.jl [hunk 3]
```diff
@@ -709,7 +709,7 @@ function h5read(filename, name::String; pv...)
     local dat
     fid = h5open(filename, "r"; pv...)
     try
-        obj = fid[name, pv...]
```
Comment (musm) [inline src/HDF5.jl:713]: Could we split this out to a separate PR  ?

Comment (musm) [reply src/HDF5.jl:713]: cc @kleinhenz  I think we missed this in the big update

Comment (jipolanco) [reply src/HDF5.jl:713]: I thought about that... I agree.

**Comment by jipolanco** [comment]
Well, the parallel HDF5 tests (`test/mpio.jl`) are not run in CI, because they require MPI-enabled HDF5 libraries. This is not the case for the default HDF5 libraries provided by BinaryBuilder. It is however possible to test parallel HDF5 on Travis (I have a package where [this is done](https://github.com/jipolanco/PencilFFTs.jl/blob/master/.travis.yml)).

The proposed interface is already easier to use, but I think we can do better. For the moment I tried to follow the HDF5 C API (e.g. by explicitly passing `*_mpio` property lists), but it's true that we can take inspiration from h5py. For instance, it would make sense to have a `h5open(filename, [mode], comm::MPI.Comm, [info::MPI.Info])` method that would create a parallel HDF5 file. We could ask @simonbyrne to review this, if he's interested.

As for the documentation, I already wrote a bit on parallel HDF5 in the [docs for my PencilFFTs.jl package](https://jipolanco.github.io/PencilFFTs.jl/latest/PencilIO/), which wraps the existent HDF5.jl interface. I think it would make sense to transfer part of those docs to HDF5.jl (and I'd have no problem with doing that).

File: src/HDF5.jl (hunk 3 above)
Comment (jipolanco) [reply src/HDF5.jl:713]: Done, see #651

**Comment by musm** [comment]
Looking at this again, it is certainly an improvement even if I'm not familiar with the MPI code, I can tell that. 

How close do you think this PR is? Are we just waiting on documentation?

**Comment by jipolanco** [comment]
Sorry for the delay, I was a bit busy with other things.

I just added a `h5open` variant for opening files in parallel more conveniently, as I proposed in my previous post.

I also added a section on parallel HDF5 to the docs. It includes a guide on how to set-up parallel HDF5, as well as a small example. To make things more consistent, I also copied a section of the README on how to use custom HDF5 libraries (which is needed for parallel HDF5) into the docs.

Personally now I'm happy with the changes, and I'd say this is ready to be merged.

File: src/mpio.jl [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+Equivalent to `h5open(filename, mode; fapl_mpio=(comm, info), pv...)`.
+Throws an informative error if the loaded HDF5 libraries do not include parallel
+support.
+
+See the [HDF5 docs](https://portal.hdfgroup.org/display/HDF5/H5P_SET_FAPL_MPIO)
+for details on the `comm` and `info` arguments.
+"""
+function h5open(
+        filename::AbstractString, mode::AbstractString,
+        comm::MPI.Comm, info::MPI.Info = MPI.Info(); pv...
```
Comment (musm) [inline src/mpio.jl:87]: I'm thinking these should actually be keyword arguments.  Would `driver` be a more appropriate name for `info` ?

File: docs/src/index.md [hunk 5]
```diff
@@ -362,6 +362,116 @@ HDF5 is a large library, and the low-level wrap is not complete. However, many o
...
+myrank = MPI.Comm_rank(comm)
+M = 10
+A = fill(myrank, M)  # local data
+dims = (M, Nproc)    # dimensions of global data
+
+# Create dataset
+dset = d_create(ff, "/data", datatype(eltype(A)), dataspace(dims))
+
+# Write local data
+dset[:, myrank + 1] = A
```
Comment (musm) [inline docs/src/index.md:456]: Why is ` +1 ` always passed when dealing with rank? Should the library automatically add think?

**Comment by musm** [comment]
Great thanks, docs look great!

File: src/mpio.jl (hunk 4 above)
Comment (jipolanco) [reply src/mpio.jl:87]: I guess you're thinking about the h5py interface. Note that the `driver` argument in h5py is different from the `info` argument here, which [allows to pass "hints"](https://portal.hdfgroup.org/display/knowledge/How+to+pass+hints+to+MPI+from+HDF5) to the MPI implementation. This can be very important for I/O performance in large computing clusters. I'm a bit surprised that h5py doesn't support this (or maybe it does, but it's not mentioned in the docs).

In my opinion, having a `driver` argument is redundant. If the user passes MPI objects to `h5open`, then it should be clear that they want to perform parallel I/O. I also like the fact that we can take advantage of multiple dispatch to provide a `h5open` variant taking MPI objects, without the need for keyword arguments.

File: docs/src/index.md (hunk 5 above)
Comment (jipolanco) [reply docs/src/index.md:456]: That's an issue coming from the MPI side, and I think there's not much we can do about it.

In MPI.jl, the process ranks returned by `MPI.Comm_rank` are in `0:Nproc-1`. Although this is not the most natural choice in Julia, it is consistent with the MPI C and Fortran APIs, and MPI.jl users are usually aware of this.

From the HDF5 side, we can't assume that the index passed to `setindex!` is an MPI rank, as it could be anything else.

File: src/mpio.jl (hunk 4 above)
Comment (musm) [reply src/mpio.jl:87]: Yeah I was. 

Regarding the arguments, should these be keyword arguments instead of positional. I feel like they should be.

Comment (jipolanco) [reply src/mpio.jl:87]: In my opinion they should be positional arguments, but I guess that's just personal preference. I'd say it's more natural in Julia, as it takes advantage of multiple dispatch to enable different functionalities.

More practically, an advantage of using positional arguments is that Julia doesn't perform dispatch based on keyword arguments. Therefore, if we were to provide a `h5open(args...; comm, info, etc...)` method, then the new keyword arguments `comm` and `info` would need to be added to the main `h5open` variants in `src/HDF5.jl`. These variants are not aware of MPI, so we cannot set e.g. `info = MPI.Info()` as default value, and we would need to set default values such as `comm = nothing, info = nothing`. In the end, I think this would just increase complexity and visual noise in the main HDF5.jl file, and would go against the idea of putting all MPI-related code in a separate file.

Comment (musm) [reply src/mpio.jl:87]: Thanks for the clarification. So the main argument for keeping them positional,  is that default values can be passed with values that depend on `MPI`  as their defaults, and we obviously can't have these as default values if we made them keyword arguments.

Comment (musm) [reply src/mpio.jl:87]: I guess keyword based dispatch (a la https://github.com/simonbyrne/KeywordDispatch.jl )  would help since that would free the keywords `comm, info` since they would dispatch on `MPI` types, but it also seems perhaps out of scope.

**Comment by musm** [comment]
Thanks again, and sorry this took so long to review and merge. Keep bumping until you force me to do it next time 😄

**Comment by jipolanco** [comment]
Great, thanks for merging!
